Infinite loop upon encode/1 with inject: true

koudelka opened the pull request:
When using the `inject` option, it looks like the `defs` function (https://github.com/bitwalker/exprotobuf/blob/master/lib/exprotobuf/define_message.ex#L87 ) loops infinitely. `encode/1` calls `defs` (as an argument to `Encoder.encode/2`), `defs(_ \\ nil)` responds and calls `@root.defs` (itself)

Should injected modules also get the global defs helper from https://github.com/bitwalker/exprotobuf/blob/master/lib/exprotobuf/builder.ex#L77?

Test case below:

```elixir
defmodule Basic do
  use Protobuf, from: Path.expand("../test/proto/basic.proto", __DIR__), only: :Basic, inject: true
end
```

```
Interactive Elixir (1.0.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> Basic.new
%Basic{args: nil, f1: nil, f2: nil, type: nil}
iex(2)> Basic.new |> Basic.encode

BREAK: (a)bort (c)ontinue (p)roc info (i)nfo (l)oaded
       (v)ersion (k)ill (D)b-tables (d)istribution
^C%
```

koudelka commented:
this fix may have introduced some regressions upon attempting to use it in a more complex case.

bitwalker commented:
@koudelka I'm out of the country until the 5th of April, so I don't have a good chance to review this currently. Have you addressed the regressions you found in your initial PR? Is this ready to be merged, or should I hold off until I can review it further?

koudelka commented:
Hey there, sorry for the delay.

I've fixed the regression that I created. As you mentioned in the README, the `inject` option is a bit touchy at runtime.

I've added the ability to use `inject` without `only`, the message name to be injected into the module is just taken from the module name. I think the case where `inject` and `only` are used together would probably need some larger design decisions before it could be more robustly supported, ditto for when injecting into a module that isn't exactly the name of the message. (I'm not sure if the inject feature is really used by anyone though?)

bitwalker commented:
@koudelka I don't know how I let this slide past me, but could you rebase your changes on master and resubmit? I'll merge once you're ready.

koudelka commented:
@bitwalker Sorry for my reciprocal delay. I've rebased my patch off the current `master` and fixed it up. :)

bitwalker commented:
Awesome :), thanks for following up!

bitwalker commented:
Released in 0.10.2 for those interested
